x-pack/plugins/lens/public/xy_visualization_plugin/xy_expression.tsx

```diff
@@ -0,0 +1,245 @@
+ */
+const axisConfig = {
+  title: { types: ['string'] },
+  showGridlines: { types: ['boolean'] },
+  position: { types: ['string'] },
+};
+
+/**
+ * Y-axis configuration.
+ */
+export interface YConfig extends AxisConfig {
+  accessors: string[];
+}
+
+/**
+ * YConfig as an expression type.
+ */
+export const yConfig = {
+  name: 'yConfig',
+  aliases: [],
+  type: 'yConfig',
+  help: `Configure a chart's y axis`,
+  context: {
+    types: ['null'],
+  },
+  args: {
+    ...axisConfig,
+    accessors: {
+      types: ['string'],
+      multi: true,
+    },
+  },
+  fn: function fn(_context: any, args: YConfig) {
+    return {
+      type: 'yConfig',
+      ...args,
+    };
+  },
+};
+
+/**
+ * X-axis configuration.
+ */
+export interface XConfig extends AxisConfig {
+  accessor: string;
+}
+
+/**
+ * XConfig as an expression type.
+ */
+export const xConfig = {
+  name: 'xConfig',
+  aliases: [],
+  type: 'xConfig',
+  help: `Configure a chart's x axis`,
+  context: {
+    types: ['null'],
+  },
+  args: {
+    ...axisConfig,
+    accessor: { types: ['string'] },
+  },
+  fn: function fn(_context: any, args: XConfig) {
+    return {
+      type: 'xConfig',
+      ...args,
+    };
+  },
+};
+
+/**
+ * The arguments to the XY chart expression function.
+ */
+export interface XYArgs {
+  seriesType: 'bar' | 'line' | 'area';
+  title: string;
+  legend: LegendConfig;
+  y: YConfig;
+  x: XConfig;
+  splitSeriesAccessors: string[];
+  stackAccessors: string[];
+}
+
+/**
+ * The XY chart expression function.
+ */
+export const xyChart = {
+  name: 'xy_chart',
+  type: 'render',
+  args: {
+    seriesType: { types: ['string'] },
+    title: { types: ['string'] },
+    legend: { types: ['legendConfig'] },
+    y: { types: ['yConfig'] },
+    x: { types: ['xConfig'] },
+    splitSeriesAccessors: {
+      types: ['string'],
+      multi: true,
+    },
+    stackAccessors: {
+      types: ['string'],
+      multi: true,
+    },
+  },
+  context: { types: ['kibana_datatable'] },
+  fn(data: Datatable, args: XYArgs) {
+    return {
+      type: 'render',
+      as: 'xy_chart_renderer',
+      value: {
+        data,
+        args,
+      },
+    };
+  },
+};
+
+/**
+ * The XY chart expression renderer.
+ */
+export const xyChartRenderer = {
+  name: 'xy_chart_renderer',
+  displayName: 'XY Chart',
+  reuseDomNode: true,
+  render: async (domNode: HTMLDivElement, config: any, _handlers: any) => {
+    ReactDOM.render(<XYChart {...config} />, domNode);
+  },
+};
+
+/**
+ * Render an XY chart as a React component.
+ *
+ * @param props - The data and args which are used to render the chart.
+ */
+function XYChart({ data, args }: { data: Datatable; args: XYArgs }) {
+  const { legend, x, y, splitSeriesAccessors, stackAccessors, seriesType } = args;
+  const seriesProps = {
+    splitSeriesAccessors,
+    stackAccessors,
+    id: getSpecId(y.accessors.join(',')),
+    xAccessor: x.accessor,
+    yAccessors: y.accessors,
+    data: data.rows,
```

Comment:
Not sure if you've seen this case yet, but elastic-charts requires some preprocessing of data such as for date types.

Reply:
Yeap. That's all stuff that I think will be easier to hammer out when we actually have this rendering. It's kind of hard to nail things like that down until then. Most of this PR was tested in a standalone project, where I had a formatter for each axis. Formatters are something we need to support, but I think it makes sense as its own PR.